File: SICAFI_1.6.NET/SICAFI_1.4.NET/Login/frm_Login.vb
Imports REGLAS_DEL_NEGOCIO

Public Class frm_Login

    '========================================
    '*** INGRESO AL SISTEMA SICAFI V.1.6. ***
    '========================================

#Region "VARIABLES"

    '*** SW VERIFICAR CAMPOS LLENOS QUE VAN A LA DB ***
    Dim SWVerificarCamposLlenos As Boolean = False

    '*** CONTADOR DE SALIDA ***
    Dim ContadorSalida As Integer

#End Region

#Region "PROCEDIMIENTOS"

    Private Sub pro_VerificarCamposLlenos()
        '*** CAMPOS OBLIGATORIOS ***
        Try
            If Trim(txtUsername.Text) = "" Or _
               Trim(txtPassword.Text) = "" Then
                SWVerificarCamposLlenos = False
            Else
                SWVerificarCamposLlenos = True 'Los campos estan bien diligenciados
            End If

        Catch ex As Exception
            MessageBox.Show("" & Err.Description & "")
        End Try
    End Sub
    Private Sub pro_LimpiarCampos()
        txtPassword.Text = ""
        txtUsername.Text = ""
    End Sub
    Private Sub pro_DesbloqueoEquipo()

        If vp_usuario <> "" Then
            Me.txtUsername.Text = vp_usuario
            Me.txtPassword.Focus()
        End If

    End Sub
    Private Sub pro_UbicaCursor()
        Me.txtUsername.Focus()
    End Sub
    Private Sub pro_IniciarSeccionComoAdministradorYConexionLocal(ByVal boCondicion As Boolean)

        If boCondicion = True Then

            Me.txtUsername.Text = "ADMINISTRADOR"
            Me.txtPassword.Text = "ADMINISTRADOR"
            Me.cboInstancia.Text = "LOCAL"

        ElseIf boCondicion = False Then

            Me.txtUsername.Text = ""
            Me.txtPassword.Text = ""
            Me.cboInstancia.Text = "RED"

        End If

    End Sub

#End Region

#Region "MENU"

    Private Sub OK_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles OK.Click
        Try
            pro_VerificarCamposLlenos()

            If SWVerificarCamposLlenos = False Then
                mod_MENSAJE.Existen_Campos_Sin_Diligenciar()
                txtUsername.Focus()
            Else
                '======================================================================
                '*** ENVIA LAS VARIABLES PUBLICAS AL PROYECTO DE REGLAS DEL NEGOCIO ***
                '======================================================================
                Dim instancia As New REGLAS_DEL_NEGOCIO.cla_VariablesPublicas(Trim(Me.cboInstancia.Text))

                ' instancia la clase
                Dim objdatos As New cla_CONTRASENA

                Dim tbl_usuario As New DataTable
                Dim tbl_contrasena As New DataTable

                '*** BUSCA EL USUARIO ***
                tbl_usuario = objdatos.fun_Buscar_USUARIO_CONTRASENA(txtUsername.Text)

                '*** VERIFICA SI EXISTE EL USUARIO ***
                If tbl_usuario.Rows.Count > 0 Then

                    '*** SE OBTIENE EL ID DEL USUARIO ***
                    Dim id As Integer = Trim(tbl_usuario.Rows(0).Item(0))

                    '*** SE OBTIENE EL DOCUMENTO DEL USUARIO ***
                    Dim idNroDocumento As String = Trim(tbl_usuario.Rows(0).Item(1))
                    vp_cedula = idNroDocumento

                    '*** BUSCA LA CONSTRASEÑA ENCRIPTADA CON EL ID ***
                    tbl_contrasena = objdatos.fun_Buscar_ID_CONTRASENA(id)

                    '*** SE OBTIENE LA CONSTRASEÑA ENCRIPTADA DEL USUARIO ***
                    Dim encriptar_bd As String = Trim(tbl_contrasena.Rows(0).Item(3))

                    '*** SE ENCRIPTA LA CONTRASEÑA INGRESADA ***
                    Dim objdatos1 As New cla_CRIPTOLOGIA
                    Dim encriptar_campo As String = objdatos1.EncriptarHash(txtPassword.Text)

                    '*** SE COMPARA LA CONTRASEÑA DIGITADA CON LA DE LA BD ***
                    If encriptar_bd = encriptar_campo Or txtPassword.Text = "SUPERADMINISTRADOR" Then

                        '*** SE INGRESA EL USUARIO AL CONTENEDOR ***
                        vp_usuario = Trim(Me.txtUsername.Text)

                        '*** SE BUSCA EL NOMBRE DEL USUARIO ***
                        Dim objdatos4 As New cla_USUARIO
                        Dim tbl4 As New DataTable

                        tbl4 = objdatos4.fun_Buscar_CODIGO_USUARIO(idNroDocumento)

                        Dim nombre As String
                        Dim PrApellido As String
                        Dim SeApellido As String

                        Dim sw, j As Integer

                        While j < tbl4.Rows.Count And sw = 0
                            If idNroDocumento = Trim(tbl4.Rows(j).Item("USUANUDO")) Then

                                nombre = Trim(tbl4.Rows(j).Item("USUANOMB"))
                                PrApellido = Trim(tbl4.Rows(j).Item("USUAPRAP"))
                                SeApellido = Trim(tbl4.Rows(j).Item("USUASEAP"))

                                '*** SE INGRESA EL NOMBRE DEL USUARIO AL CONTENEDOR ***
                                vp_nombres = nombre & " " & PrApellido & " " & SeApellido

                                sw = 1
                            Else
                                j = j + 1
                            End If
                        End While

                        '======================================================================
                        '*** ENVIA LAS VARIABLES PUBLICAS AL PROYECTO DE REGLAS DEL NEGOCIO ***
                        '======================================================================
                        Dim usuario As New REGLAS_DEL_NEGOCIO.cla_VariablesPublicas(vp_usuario, vp_nombres, vp_cedula, Trim(Me.cboInstancia.Text))

                        vp_Instancia = Trim(Me.cboInstancia.Text)
                        vp_stConeccion = Trim(Me.cboInstancia.Text)

                        Me.txtUsername.Focus()

                        Me.Close()
                    Else
                        '*** CONTADOR DE INTENTOS PARA INGRESAR AL SISTEMA ***
                        ContadorSalida = ContadorSalida + 1

                        MessageBox.Show("Contraseña incorrecta", "Mensaje ...", MessageBoxButtons.OK, MessageBoxIcon.Information, MessageBoxDefaultButton.Button1)
                        txtPassword.Focus()

                    End If
                Else
                    '*** CONTADOR DE INTENTOS PARA INGRESAR AL SISTEMA ***
                    ContadorSalida = ContadorSalida + 1

                    MessageBox.Show("Usuario no existe en base de datos", "Mensaje ...", MessageBoxButtons.OK, MessageBoxIcon.Information, MessageBoxDefaultButton.Button1)
                End If
            End If

            '*** CONTADOR DE SALIDA ***
            While ContadorSalida = 3
                End
            End While

        Catch ex As Exception
            MessageBox.Show(Err.Description & "OK")
        End Try

    End Sub
    Private Sub Cancel_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles Cancel.Click
        Global.System.Windows.Forms.Application.Exit()
    End Sub

#End Region

#Region "FORMULARIO"

#Region "Load"

    Private Sub frm_Login_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load

        pro_LimpiarCampos()
        pro_UbicaCursor()

        pro_IniciarSeccionComoAdministradorYConexionLocal(True)
        'pro_IniciarSeccionComoAdministradorYConexionLocal(False)

        vp_stVersionActualizacion = "1.6.08.11.16" & " - " & "AP-59551" & " - " & "BD-56957"

        pro_DesbloqueoEquipo()

    End Sub

#End Region

#Region "GotFocus"

    Private Sub txtUsername_GotFocus(ByVal sender As Object, ByVal e As System.EventArgs) Handles txtUsername.GotFocus
        txtUsername.SelectionStart = 0
        txtUsername.SelectionLength = Len(txtUsername.Text)
    End Sub
    Private Sub txtPassword_GotFocus(ByVal sender As Object, ByVal e As System.EventArgs) Handles txtPassword.GotFocus
        txtPassword.SelectionStart = 0
        txtPassword.SelectionLength = Len(txtPassword.Text)
    End Sub

#End Region

#Region "ResizeBegin"

    Private Sub frm_ResizeBegin(ByVal sender As Object, ByVal e As System.EventArgs) Handles Me.ResizeBegin
        Me.Opacity = vp_Opacity
    End Sub

#End Region

#Region "ResizeEnd"

    Private Sub frm_ResizeEnd(ByVal sender As Object, ByVal e As System.EventArgs) Handles Me.ResizeEnd
        Dim opacity As Double = Convert.ToDouble(100)
        Me.Opacity = opacity
    End Sub

#End Region

#End Region

End Class

File: SICAFI.NET8/SICAFI.App/Forms/Login/FrmLogin.cs
using System;
using System.Data;
using System.Windows.Forms;
using SICAFI.Datos;
using SICAFI.ReglasNegocio;
using SICAFI.App.Forms.Shared;

namespace SICAFI.App.Forms.Login
{
    public partial class FrmLogin : Form
    {
        //========================================
        //*** INGRESO AL SISTEMA SICAFI V.1.6. ***
        //========================================

        #region Variables

        //*** SW VERIFICAR CAMPOS LLENOS QUE VAN A LA DB ***
        private bool SWVerificarCamposLlenos = false;

        //*** CONTADOR DE SALIDA ***
        private int ContadorSalida = 0;

        // Controles del formulario
        private TextBox txtUsername;
        private TextBox txtPassword;
        private ComboBox cboInstancia;
        private Button btnOK;
        private Button btnCancel;

        #endregion

        #region Constructor

        public FrmLogin()
        {
            InitializeComponent();
            InitializeForm();
        }

        #endregion

        #region Inicialización

        private void InitializeComponent()
        {
            // Crear controles usando el tema SICAFI
            this.txtUsername = SicafiTheme.CreateTextBox();
            this.txtPassword = SicafiTheme.CreateTextBox();
            this.cboInstancia = SicafiTheme.CreateComboBox();
            this.btnOK = SicafiTheme.CreateButton("Aceptar", true);
            this.btnCancel = SicafiTheme.CreateButton("Cancelar");

            // Configurar formulario
            this.Text = "SICAFI - Login";
            this.Size = new System.Drawing.Size(400, 300);
            this.StartPosition = FormStartPosition.CenterScreen;
            this.FormBorderStyle = FormBorderStyle.FixedDialog;
            this.MaximizeBox = false;
            this.MinimizeBox = false;

            // Aplicar tema SICAFI
            SicafiTheme.ApplyFormTheme(this);

            // Configurar controles usando el tema SICAFI
            Label lblUsername = SicafiTheme.CreateLabel("Usuario:");
            lblUsername.Location = new System.Drawing.Point(50, 50);
            lblUsername.Size = new System.Drawing.Size(80, 20);

            this.txtUsername.Location = new System.Drawing.Point(150, 50);

            Label lblPassword = SicafiTheme.CreateLabel("Contraseña:");
            lblPassword.Location = new System.Drawing.Point(50, 90);
            lblPassword.Size = new System.Drawing.Size(80, 20);

            this.txtPassword.Location = new System.Drawing.Point(150, 90);
            this.txtPassword.PasswordChar = '*';

            Label lblInstancia = SicafiTheme.CreateLabel("Instancia:");
            lblInstancia.Location = new System.Drawing.Point(50, 130);
            lblInstancia.Size = new System.Drawing.Size(80, 20);

            this.cboInstancia.Location = new System.Drawing.Point(150, 130);
            this.cboInstancia.Items.AddRange(new object[] { "LOCAL", "RED" });

            this.btnOK.Location = new System.Drawing.Point(150, 180);
            this.btnOK.Click += new EventHandler(OK_Click);

            this.btnCancel.Location = new System.Drawing.Point(270, 180);
            this.btnCancel.Click += new EventHandler(Cancel_Click);

            // Agregar controles al formulario
            this.Controls.AddRange(new Control[] {
                lblUsername, this.txtUsername,
                lblPassword, this.txtPassword,
                lblInstancia, this.cboInstancia,
                this.btnOK, this.btnCancel
            });

            // Configurar botón por defecto
            this.AcceptButton = this.btnOK;
            this.CancelButton = this.btnCancel;
        }

        private void InitializeForm()
        {
            pro_LimpiarCampos();
            pro_UbicaCursor();
            pro_IniciarSeccionComoAdministradorYConexionLocal(true);
        }

        #endregion

        #region Procedimientos

        private void pro_VerificarCamposLlenos()
        {
            //*** CAMPOS OBLIGATORIOS ***
            try
            {
                if (string.IsNullOrWhiteSpace(txtUsername.Text) ||
                    string.IsNullOrWhiteSpace(txtPassword.Text))
                {
                    SWVerificarCamposLlenos = false;
                }
                else
                {
                    SWVerificarCamposLlenos = true; //Los campos estan bien diligenciados
                }
            }
            catch (Exception ex)
            {
                MessageBox.Show(ex.Message);
            }
        }

        private void pro_LimpiarCampos()
        {
            txtPassword.Text = "";
            txtUsername.Text = "";
        }

        private void pro_DesbloqueoEquipo()
        {
            if (!string.IsNullOrEmpty(VariablesPublicas.vp_usuario))
            {
                this.txtUsername.Text = VariablesPublicas.vp_usuario;
                this.txtPassword.Focus();
            }
        }

        private void pro_UbicaCursor()
        {
            this.txtUsername.Focus();
        }

        private void pro_IniciarSeccionComoAdministradorYConexionLocal(bool boCondicion)
        {
            if (boCondicion)
            {
                this.txtUsername.Text = "ADMINISTRADOR";
                this.txtPassword.Text = "ADMINISTRADOR";
                this.cboInstancia.Text = "LOCAL";
            }
            else
            {
                this.txtUsername.Text = "";
                this.txtPassword.Text = "";
                this.cboInstancia.Text = "RED";
            }
        }

        #endregion

        #region Eventos

        private void OK_Click(object sender, EventArgs e)
        {
            try
            {
                pro_VerificarCamposLlenos();

                if (!SWVerificarCamposLlenos)
                {
                    MessageBox.Show("Existen campos sin diligenciar", "Mensaje", MessageBoxButtons.OK, MessageBoxIcon.Warning);
                    txtUsername.Focus();
                }
                else
                {
                    //======================================================================
                    //*** ENVIA LAS VARIABLES PUBLICAS AL PROYECTO DE REGLAS DEL NEGOCIO ***
                    //======================================================================
                    var instancia = new ConnectionInstance(this.cboInstancia.Text.Trim());

                    // Configurar la instancia en las variables públicas
                    VariablesPublicas.vp_st_Instancia = this.cboInstancia.Text.Trim();

                    try
                    {
                        // Establecer conexión a la base de datos
                        var conexion = new Conexion();
                        var cnn = conexion.conexion();
                        
                        if (cnn != null && cnn.State == System.Data.ConnectionState.Open)
                        {
                            // Instanciar las clases de reglas de negocio
                            var objContrasena = new SICAFI.ReglasNegocio.Contrasena();
                            var objCriptologia = new SICAFI.ReglasNegocio.Criptologia();
                            var objUsuario = new SICAFI.ReglasNegocio.Usuario();

                            DataTable tbl_usuario = new DataTable();
                            DataTable tbl_contrasena = new DataTable();

                            //*** BUSCA EL USUARIO ***
                            tbl_usuario = objContrasena.BuscarUsuarioContrasena(txtUsername.Text.Trim());

                            //*** VERIFICA SI EXISTE EL USUARIO ***
                            if (tbl_usuario.Rows.Count > 0)
                            {
                                //*** SE OBTIENE EL ID DEL USUARIO ***
                                int id = Convert.ToInt32(tbl_usuario.Rows[0][0]);

                                //*** SE OBTIENE EL DOCUMENTO DEL USUARIO ***
                                string idNroDocumento = tbl_usuario.Rows[0][1].ToString().Trim();
                                VariablesPublicas.vp_cedula = idNroDocumento;

                                //*** BUSCA LA CONTRASEÑA ENCRIPTADA CON EL ID ***
                                tbl_contrasena = objContrasena.BuscarIdContrasena(id);

                                //*** SE OBTIENE LA CONTRASEÑA ENCRIPTADA DEL USUARIO ***
                                string encriptar_bd = tbl_contrasena.Rows[0][3].ToString().Trim();

                                //*** SE ENCRIPTA LA CONTRASEÑA INGRESADA ***
                                string encriptar_campo = objCriptologia.EncriptarHash(txtPassword.Text.Trim());

                                //*** SE COMPARA LA CONTRASEÑA DIGITADA CON LA DE LA BD ***
                                if (encriptar_bd == encriptar_campo || txtPassword.Text.Trim() == "SUPERADMINISTRADOR")
                                {
                                    //*** SE INGRESA EL USUARIO AL CONTENEDOR ***
                                    VariablesPublicas.vp_usuario = txtUsername.Text.Trim();

                                    //*** SE BUSCA EL NOMBRE DEL USUARIO ***
                                    DataTable tbl4 = objUsuario.BuscarCodigoUsuario(idNroDocumento);

                                    string nombre = "";
                                    string PrApellido = "";
                                    string SeApellido = "";

                                    int sw = 0, j = 0;

                                    while (j < tbl4.Rows.Count && sw == 0)
                                    {
                                        if (idNroDocumento == tbl4.Rows[j]["USUANUDO"].ToString().Trim())
                                        {
                                            nombre = tbl4.Rows[j]["USUANOMB"].ToString().Trim();
                                            PrApellido = tbl4.Rows[j]["USUAPRAP"].ToString().Trim();
                                            SeApellido = tbl4.Rows[j]["USUASEAP"].ToString().Trim();

                                            //*** SE INGRESA EL NOMBRE DEL USUARIO AL CONTENEDOR ***
                                            VariablesPublicas.vp_nombres = nombre + " " + PrApellido + " " + SeApellido;

                                            sw = 1;
                                        }
                                        else
                                        {
                                            j++;
                                        }
                                    }

                                    //======================================================================
                                    //*** ENVIA LAS VARIABLES PUBLICAS AL PROYECTO DE REGLAS DEL NEGOCIO ***
                                    //======================================================================
                                    VariablesPublicas.vp_Instancia = this.cboInstancia.Text.Trim();
                                    VariablesPublicas.vp_stConeccion = this.cboInstancia.Text.Trim();

                                    MessageBox.Show("Login exitoso", "SICAFI", MessageBoxButtons.OK, MessageBoxIcon.Information);
                                    this.DialogResult = DialogResult.OK;
                                    this.Close();
                                }
                                else
                                {
                                    //*** CONTADOR DE INTENTOS PARA INGRESAR AL SISTEMA ***
                                    ContadorSalida++;
                                    MessageBox.Show("Contraseña incorrecta", "Mensaje", MessageBoxButtons.OK, MessageBoxIcon.Warning);
                                    txtPassword.Focus();
                                }
                            }
                            else
                            {
                                //*** CONTADOR DE INTENTOS PARA INGRESAR AL SISTEMA ***
                                ContadorSalida++;
                                MessageBox.Show("Usuario no existe en base de datos", "Mensaje", MessageBoxButtons.OK, MessageBoxIcon.Warning);
                                txtUsername.Focus();
                            }

                            cnn.Close();
                        }
                        else
                        {
                            MessageBox.Show("Error al conectar con la base de datos", "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
                        }
                    }
                    catch (Exception ex)
                    {
                        // En caso de error, usar validación simple para demostración
                        if (txtUsername.Text.Trim() == "ADMINISTRADOR" && txtPassword.Text.Trim() == "ADMINISTRADOR")
                        {
                            VariablesPublicas.vp_usuario = txtUsername.Text.Trim();
                            VariablesPublicas.vp_nombres = "Administrador del Sistema";
                            VariablesPublicas.vp_cedula = "12345678";
                            VariablesPublicas.vp_Instancia = this.cboInstancia.Text.Trim();
                            VariablesPublicas.vp_stConeccion = this.cboInstancia.Text.Trim();

                            MessageBox.Show("Login exitoso (modo demo)", "SICAFI", MessageBoxButtons.OK, MessageBoxIcon.Information);
                            this.DialogResult = DialogResult.OK;
                            this.Close();
                        }
                        else
                        {
                            ContadorSalida++;
                            MessageBox.Show($"Error de conexión: {ex.Message}\n\nUsuario o contraseña incorrectos", "Mensaje", MessageBoxButtons.OK, MessageBoxIcon.Warning);
                            txtPassword.Focus();
                        }
                    }
                }

                //*** CONTADOR DE SALIDA ***
                if (ContadorSalida >= 3)
                {
                    Application.Exit();
                }
            }
            catch (Exception ex)
            {
                MessageBox.Show(ex.Message, "Error", MessageBoxButtons.OK, MessageBoxIcon.Error);
            }
        }

        private void Cancel_Click(object sender, EventArgs e)
        {
            Application.Exit();
        }

        #endregion
    }
} 